controllers/topology/internal/extensions/patches/inline/json_patch_generator.go

```diff
@@ -0,0 +1,372 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package inline implements the inline JSON patch generator.
+package inline
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"text/template"
+
+	"github.com/pkg/errors"
+	"github.com/valyala/fastjson"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/extensions/patches/api"
+	patchvariables "sigs.k8s.io/cluster-api/controllers/topology/internal/extensions/patches/variables"
+	"sigs.k8s.io/yaml"
+)
+
+// JSONPatchGenerator generates JSON patches for a GenerateRequest based on a ClusterClassPatch.
+type JSONPatchGenerator struct {
+	patch *clusterv1.ClusterClassPatch
+}
+
+// NewJSONPatchGenerator returns a JSONPatchGenerator from a given ClusterClassPatch object.
+func NewJSONPatchGenerator(patch *clusterv1.ClusterClassPatch) api.Generator {
+	return &JSONPatchGenerator{
+		patch: patch,
+	}
+}
+
+// Generate generates JSON patches for the given GenerateRequest based on a ClusterClassPatch.
+func (j *JSONPatchGenerator) Generate(_ context.Context, req *api.GenerateRequest) (*api.GenerateResponse, error) {
+	resp := &api.GenerateResponse{}
+
+	// Loop over all templates.
+	for _, template := range req.Items {
+		// Calculate the list of patches which match the current template.
+		matchingPatches := []clusterv1.PatchDefinition{}
+		for _, patch := range j.patch.Definitions {
+			// Add the patch to the list, if it matches the template.
+			if templateMatchesSelector(&template.TemplateRef, patch.Selector) {
+				matchingPatches = append(matchingPatches, patch)
+			}
+		}
+
+		// Continue if there are no matching patches.
+		if len(matchingPatches) == 0 {
+			continue
+		}
+
+		// Merge template-specific and global variables.
+		// NOTE: To optimize performance, we only do this once for every template.
+		variables, err := mergeVariableMaps(req.Variables, template.Variables)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to merge global and template-specific variables")
+		}
+
+		// Loop over all PatchDefinitions.
+		for _, patch := range matchingPatches {
+			// Generate JSON patches.
+			jsonPatches, err := generateJSONPatches(patch.JSONPatches, variables)
+			if err != nil {
+				return nil, err
+			}
```

Comment:
In debugging patches would be helpful to aggregate these errors and then returning them to the caller, have we considered to collect them instead of returning right away?

Reply:
Sounds like a good idea.

Reply:
Done